File: advent_of_code/03.go
package main

import (
	"fmt"
)

func _03() {
	lines := getLines("03.txt")

	getPriority := func(item byte) int {
		i := int(item)
		if i > int('a') {
			return i - int('a') + 1
		} else {
			return i - int('A') + 27
		}
	}

	part1 := func() {
		s := 0
		for _, line := range lines {
			halfSize := len(line) / 2
			firstHalf := make(map[byte]bool)
			for c := 0; c < len(line); c++ {
				item := line[c]
				if c < halfSize {
					firstHalf[item] = true
				} else if firstHalf[item] {
					s += getPriority(item)
					break
				}
			}

		}
		fmt.Println(s)
	}

	part2 := func() {
		s := 0
		var seen map[rune]bool
		for i, line := range lines {
			if i%3 == 0 {
				seen = make(map[rune]bool)
			}
			for _, item := range line {
				if i%3 == 0 {
					seen[item] = false
				} else if i%3 == 1 {
					if _, ok := seen[item]; ok {
						seen[item] = true
					}
				} else {
					if seen[item] {
						s += getPriority(byte(item))
						break
					}
				}
			}
		}
		fmt.Println(s)
	}

	part1()
	part2()

}

File: advent_of_code/2023/03.py
from collections import defaultdict

with open('03.txt') as fh:
    txt = fh.read()

digits = '1234567890'
lines = txt.split("\n")

def part1():
    def is_symbol(i, j, len_chars):
        for ii in [-1, 0, 1]:
            xi = i+ii
            if xi < 0 or xi >= len(lines):
                continue
            line = lines[xi]
            for ij in range(len_chars+2):
                xj = j - ij
                if xj < 0 or xj >= len(line):
                    continue
                char = line[xj]
                if char not in digits and char != '.':
                    return True
        return False

    s = 0
    for i, line in enumerate(lines):
        chars = ''
        for j, char in enumerate(line+'.'):
            if char in digits:
                chars += char
            elif chars != '':
                if is_symbol(i, j, len(chars)):
                    s += int(chars)
                chars = ''
    return s

def part2():
    def get_gear(i, j, len_chars):
        for ii in [-1, 0, 1]:
            xi = i+ii
            if xi < 0 or xi >= len(lines):
                continue
            line = lines[xi]
            for ij in range(len_chars+2):
                xj = j - ij
                if xj < 0 or xj >= len(line):
                    continue
                char = line[xj]
                if char == '*':
                    return str([xi, xj])
        return False


    gears = defaultdict(lambda: [])
    for i, line in enumerate(lines):
        chars = ''
        for j, char in enumerate(line+'.'):
            if char in digits:
                chars += char
            elif chars != '':
                gear = get_gear(i, j, len(chars))
                if gear:
                    gears[gear] += [int(chars)]
                chars = ''

    s = 0
    for parts in gears.values():
        if len(parts) == 2:
            s += parts[0] * parts[1]
    return s

print(part2())